frontend/pages/chat/index.tsx

```diff
@@ -0,0 +1,59 @@
+import { useState, useEffect } from 'react';
+import { io } from 'socket.io-client';
+import { Button } from '@mui/material';
+
+type ChatRoom = {
+  name: string;
+  type: boolean;
+  author: string;
+  hashedPassword?: string;
+};
+
+const socket = io('http://localhost:3001/chat');
+
+// TODO: name以外も指定できるようにする
+const createChatRoom = () => {
+  const name = String(window.prompt('チャンネル名を入力してください'));
+  const room = {
+    name: name,
+    type: true,
+    author: 'admin',
+    hashedPassword: '',
+  };
+  socket.emit('room:create', room);
+  console.log('[DEBUG] room:create', room);
+};
+
+const showRooms = (rooms: ChatRoom[]) => {
+  return rooms.map((item, i) => <li key={i}>{item.name}</li>);
+};
+
+const Chat = () => {
+  const [rooms, setRooms] = useState<ChatRoom[]>([]);
+
+  // dbに保存ができたら,backendからreceiveする
+  useEffect(() => {
+    socket.on('room:created', (data: ChatRoom) => {
+      console.log('[DEBUG] room:created', data);
+      setRooms((rooms) => [...rooms, data]);
+    });
+
+    return () => {
+      socket.off('room:created');
```

Comment:
これってつけてる理由を調べてみてどっちでも良さそうだな、と思ったんですが何かつけてる理由ありますか？

Reply:
実はあります。

開発環境だとstrict modeが効いているので、複数回レンダリングされます。
useEffectにsocket.onを書いても、レンダリングされた回数イベントハンドラーが適用されるので、
offがないと、本来1回しか呼ばれてほしくないイベントが複数呼ばれることになってしまいます。

具体的な例をあげると、チャットルームを作成するたび、同じものが2回listに追加されるという挙動になるはずです。

Reply:
参考
> the listeners must be removed in the cleanup step, in order to prevent multiple event registrations

https://socket.io/how-to/use-with-react-hooks